flink-table-store-flink/flink-table-store-flink-common/src/main/java/org/apache/flink/table/store/connector/source/TableStoreSource.java

```diff
@@ -162,15 +164,47 @@ public ScanRuntimeProvider getScanRuntimeProvider(ScanContext scanContext) {
                         .withProjection(projectFields)
                         .withPredicate(predicate)
                         .withLimit(limit)
-                        .withParallelism(
-                                Options.fromMap(table.schema().options())
-                                        .get(FlinkConnectorOptions.SCAN_PARALLELISM))
+                        .withParallelism(inferParallelism(table, predicate, limit, streaming))
                         .withWatermarkStrategy(watermarkStrategy);
 
         return new TableStoreDataStreamScanProvider(
                 !streaming, env -> sourceBuilder.withEnv(env).build());
     }
 
+    private Integer inferParallelism(
+            FileStoreTable table, Predicate predicate, Long limitCount, boolean streaming) {
+        Options options = Options.fromMap(this.table.schema().options());
+        Integer parallelism = options.get(FlinkConnectorOptions.SCAN_PARALLELISM);
+
+        if (options.get(FlinkConnectorOptions.INFER_SCAN_PARALLELISM)) {
+            // for streaming mode, set the default parallelism to the bucket number.
+            if (streaming) {
+                parallelism = options.get(CoreOptions.BUCKET);
+            } else {
+                int splitSize = table.newScan().withFilter(predicate).plan().splits.size();
```

Comment:
I have concern here, plan is a heavy work, can we have a conservative approach?

Reply:
Maybe, without limit, just bucket number * 10?

Reply:
My idea is that `split` is related to the data size of table. If we use bucket size, I am worried that the data size of each bucket may be difficult to balance (if there is data skew, this situation will be more serious), so whether we can add a `totalDataSize` for snapshot,then we use `totalDataSize / 1G` to infer parallelism? Or the divisor can be specified by the user, depending on how much data the user wants each read operator to read?

In addition, the `totalDataSize` of snapshot is similar to `totalRecordCount`, which is also a metric of snapshot.

what do you think about?

Reply:
I think your thought is good~ But we should consider filter pushdown too, especially for partition&primary_key filter pushdown.

Another approach is doing some refactor. We can produce splits in compilation phase, and pass these splits to runtime source.

Reply:
> I think your thought is good~ But we should consider filter pushdown too, especially for partition&primary_key filter pushdown.


 yes, if there are query conditions, the `split` will be different

Reply:
> Another approach is doing some refactor. We can produce splits in compilation phase, and pass these splits to runtime source.

Let me investigate it

Reply:
> I think your thought is good~ But we should consider filter pushdown too, especially for partition&primary_key filter pushdown.
> 
> Another approach is doing some refactor. We can produce splits in compilation phase, and pass these splits to runtime source.

Hi,@JingsongLi, I generated the split in `FlinkTableSource`, do we need to add `snapshotId` so that we can refactor all the codes that use `scan` and `plan`,like here: https://github.com/apache/flink-table-store/blob/e715ac81a906e8f1b285d3847cc75879bb54d062/flink-table-store-flink/flink-table-store-flink-common/src/main/java/org/apache/flink/table/store/connector/source/StaticFileStoreSource.java#L86